File: app/src/main/java/org/monksanctum/xand11/comm/Event.java
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.comm;

import android.graphics.Rect;

import org.monksanctum.xand11.Client;
import org.monksanctum.xand11.ObjectPool;
import org.monksanctum.xand11.ObjectPool.Recycleable;
import org.monksanctum.xand11.Utils;

import java.lang.reflect.Field;
import java.lang.reflect.Modifier;

public class Event {
    public static final byte ERROR = 0;
    public static final byte REPLY = 1;
    public static final byte KEY_PRESS = 2;
    public static final byte KEY_RELEASE = 3;
    public static final byte BUTTON_PRESS = 4;
    public static final byte BUTTON_RELEASE = 5;

    public static final byte MOTION_NOTIFY = 6;
    public static final byte ENTER_NOTIFY = 7;
    public static final byte LEAVE_NOTIFY = 8;

    public static final byte FOCUS_IN = 9;
    public static final byte FOCUS_OUT = 10;

    public static final byte KEYMAP_NOTIFY = 11;

    public static final byte EXPOSE = 12;
    public static final byte GRAPHICS_EXPOSE = 13;
    public static final byte NO_EXPOSE = 14;
    public static final byte VISIBILITY_NOTIFY = 15;
    public static final byte CREATE_NOTIFY = 16;
    public static final byte DESTROY_NOTIFY = 17;

    public static final byte UNMAP_NOTIFY = 18;
    public static final byte MAP_NOTIFY = 19;
    public static final byte MAP_REQUEST = 20;
    public static final byte REPARENT_NOTIFY = 21;

    public static final byte CONFIGURE_NOTIFY = 22;
    public static final byte CONFIGURE_REQUEST = 23;
    public static final byte GRAVITY_NOTIFY = 24;
    public static final byte RESIZE_REQUEST = 25;
    public static final byte CIRCULATE_NOTIFY = 26;
    public static final byte CILCULATE_REQUEST = 27;

    public static final byte PROPERTY_NOTIFY = 28;
    public static final byte SELECTION_CLEAR = 29;
    public static final byte SELECTION_REQUEST = 30;
    public static final byte SELECTION_NOTIFY = 31;
    public static final byte COLORMAP_NOTIFY = 32;

    public static final byte CLIENT_MESSAGE = 33;
    public static final byte MAPPING_NOTIFY = 34;

    private static final String[] sNames = new String[256];

    // TODO: Probobly should be per client or something...?
    private static int sEvent = 1;

    public static void sendPropertyChange(Client client, int window, int atom, boolean deleted) {
        PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());

        writer.writeCard32(window);
        writer.writeCard32(atom);
        writer.writeCard32(client.getTimestamp());
        writer.writeByte((byte) (deleted ? 1 : 0));
        writer.writePadding(15);
        client.getClientListener().sendPacket(MAP_NOTIFY, writer);
    }

    public static void sendMapNotify(Client client, int window, int event,
            boolean overrideRedirect) {
        PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.writeCard32(event);
        writer.writeCard32(window);
        writer.writeByte((byte) (overrideRedirect ? 1 : 0));
        writer.writePadding(19);
        client.getClientListener().sendPacket(MAP_NOTIFY, writer);
    }

    public static void sendUnmapNotify(Client client, int window, int event,
            boolean fromConfigure) {
        PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.writeCard32(event);
        writer.writeCard32(window);
        writer.writeByte((byte) (fromConfigure ? 1 : 0));
        writer.writePadding(19);
        client.getClientListener().sendPacket(UNMAP_NOTIFY, writer);
    }

    public static void sendSelectionRequest(Client client, int timestamp, int owner, int requestor,
            int selection, int target, int property) {
        PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.writeCard32(timestamp);
        writer.writeCard32(owner);
        writer.writeCard32(requestor);
        writer.writeCard32(selection);
        writer.writeCard32(target);
        writer.writeCard32(property);
        writer.writePadding(4);
        client.getClientListener().sendPacket(SELECTION_REQUEST, writer);
    }

    public static void sendSelectionNotify(Client client, int timestamp, int requestor,
            int selection, int target, int property) {
        PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.writeCard32(timestamp);
        writer.writeCard32(requestor);
        writer.writeCard32(selection);
        writer.writeCard32(target);
        writer.writeCard32(property);
        writer.writePadding(8);
        client.getClientListener().sendPacket(SELECTION_NOTIFY, writer);
    }

    public static void sendSelectionClear(Client client, int timestamp, int owner, int selection) {
        PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.writeCard32(timestamp);
        writer.writeCard32(owner);
        writer.writeCard32(selection);
        writer.writePadding(16);
        client.getClientListener().sendPacket(SELECTION_CLEAR, writer);
    }

    public static void sendExpose(final Client client, int window, Rect bounds) {
        final PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.writeCard32(window);
        writer.writeCard16(bounds.left);
        writer.writeCard16(bounds.top);
        writer.writeCard16(bounds.width());
        writer.writeCard16(bounds.height());
        writer.writeCard16(0); // What?
        writer.writePadding(14);

        Utils.sBgHandler.post(() -> client.getClientListener().sendPacket(EXPOSE, writer));
    }

    public static void sendConfigureWindow(final Client client, int event, int window, int sibling,
            Rect bounds, int borderWidth, boolean overrideDirect) {
        final PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.writeCard32(event);
        writer.writeCard32(window);
        writer.writeCard32(sibling);
        writer.writeCard16(bounds.left);
        writer.writeCard16(bounds.top);
        writer.writeCard16(bounds.width());
        writer.writeCard16(bounds.height());
        writer.writeCard16(borderWidth);
        writer.writeByte((byte) (overrideDirect ? 1 : 0));
        writer.writePadding(5);

        client.getClientListener().sendPacket(CONFIGURE_NOTIFY, writer);
    }

    public static void sendKeyDown(Client client, int id, int x, int y, int keyCode, int state) {
        final PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.setMinorOpCode((byte) keyCode);
        writer.writeCard32(client.getTimestamp());
        writer.writeCard32(3);
        writer.writeCard32(id);
        writer.writeCard32(0); // Child
        writer.writeCard16(x); // Root-x
        writer.writeCard16(y); // Root-y
        writer.writeCard16(x);
        writer.writeCard16(y);

        writer.writeCard16(state); // Keybutton state
        writer.writeByte((byte) 1); // Same screen.
        writer.writePadding(1);

        client.getClientListener().sendPacket(KEY_PRESS, writer);
    }

    public static void sendKeyUp(Client client, int id, int x, int y, int keyCode, int state) {
        final PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.setMinorOpCode((byte) keyCode);
        writer.writeCard32(client.getTimestamp());
        writer.writeCard32(3);
        writer.writeCard32(id);
        writer.writeCard32(0); // Child
        writer.writeCard16(0); // Root-x
        writer.writeCard16(0); // Root-y
        writer.writeCard16(x);
        writer.writeCard16(y);

        writer.writeCard16(state); // Keybutton state
        writer.writeByte((byte) 1);
        writer.writePadding(1);

        client.getClientListener().sendPacket(KEY_RELEASE, writer);
    }

    public static void sendEnter(Client client, int id, int x, int y) {
        final PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.setMinorOpCode((byte) 1);

        writer.writeCard32(client.getTimestamp());
        writer.writeCard32(id);
        writer.writeCard32(id);
        writer.writeCard32(0);

        writer.writeCard16(0); // Root-x
        writer.writeCard16(0); // Root-y
        writer.writeCard16(x);
        writer.writeCard16(y);
        writer.writeCard16(0); // Keybutton state

        writer.writeByte((byte) 0); // Normal, not grab
        writer.writeByte((byte) 3); // Same screen, focus

        client.getClientListener().sendPacket(ENTER_NOTIFY, writer);
    }

    public static void sendLeave(Client client, int id, int x, int y) {
        final PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.setMinorOpCode((byte) 1);

        writer.writeCard32(client.getTimestamp());
        writer.writeCard32(id);
        writer.writeCard32(id);
        writer.writeCard32(0);

        writer.writeCard16(0); // Root-x
        writer.writeCard16(0); // Root-y
        writer.writeCard16(x);
        writer.writeCard16(y);
        writer.writeCard16(0); // Keybutton state

        writer.writeByte((byte) 0); // Normal, not grab
        writer.writeByte((byte) 3); // Same screen, focus

        client.getClientListener().sendPacket(LEAVE_NOTIFY, writer);
    }

    public static void sendDestroy(Client client, int id, int arg) {
        final PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());
        writer.writeCard32(id);
        writer.writeCard32(arg);
        writer.writePadding(20);
        client.getClientListener().sendPacket(DESTROY_NOTIFY, writer);
    }

    public static void sendCreate(Client client, int id, int child, int x, int y, int width,
            int height, int borderWidth, boolean overrideRedirect) {
        final PacketWriter writer = new PacketWriter(client.getClientListener().getWriter());

        writer.writeCard32(id);
        writer.writeCard32(child);
        writer.writeCard16(x);
        writer.writeCard16(y);
        writer.writeCard16(width);
        writer.writeCard16(height);
        writer.writeCard16(borderWidth);
        writer.writeByte((byte) (overrideRedirect ? 1 : 0));
        writer.writePadding(9);

        client.getClientListener().sendPacket(CREATE_NOTIFY, writer);
    }

    public static void populateNames() {
        Class<Event> cls = Event.class;
        for (Field field : cls.getDeclaredFields()) {
            if (Modifier.isStatic(field.getModifiers()) && field.getType() == byte.class) {
                try {
                    byte index = (byte) field.get(null);
                    sNames[index] = field.getName();
                } catch (IllegalAccessException e) {
                }
            }
        }
    }

    public static String getName(int what) {
        if (sNames[what] == null) {
            return "Unknown(" + what + ")";
        }
        return sNames[what];
    }

    public static EventInfo obtainInfo(byte type, int id, int arg) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.arg = arg;
        return e;
    }

    public static EventInfo obtainInfo(byte type, int id, int arg, boolean flag) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.arg = arg;
        e.flag = flag;
        return e;
    }

    public static EventInfo obtainInfo(byte type, int id, int arg, int arg2, int arg3, int arg4,
            int arg5, int arg6, boolean flag) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.arg = arg;
        e.arg2 = arg2;
        e.arg3 = arg3;
        e.arg4 = arg4;
        e.arg5 = arg5;
        e.arg6 = arg6;
        e.flag = flag;
        return e;
    }

    public static EventInfo obtainInfo(byte type, int id, Rect arg, boolean flag) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.rect = arg;
        e.flag = flag;
        return e;
    }

    public static EventInfo obtainInfo(byte type, int id, int arg1, int arg2, int arg3, Rect rect,
            boolean flag) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.arg = arg1;
        e.arg2 = arg2;
        e.arg3 = arg3;
        e.rect = rect;
        e.flag = flag;
        return e;
    }

    public static EventInfo obtainInfo(byte type, int id, int arg1, int arg2, int arg3, int arg4,
            int arg5) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.arg = arg1;
        e.arg2 = arg2;
        e.arg3 = arg3;
        e.arg4 = arg4;
        e.arg5 = arg5;
        return e;
    }

    public static EventInfo obtainInfo(byte type, int id, int arg1, int arg2) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.arg = arg1;
        e.arg2 = arg2;
        return e;
    }

    public static EventInfo obtainInfo(byte type, int id, int arg1, int arg2, int arg3) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.arg = arg1;
        e.arg2 = arg2;
        e.arg3 = arg3;
        return e;
    }

    public static EventInfo obtainInfo(byte type, int id, int arg1, int arg2, int arg3, int arg4) {
        EventInfo e = sEventPool.obtain();
        e.type = type;
        e.id = id;
        e.arg = arg1;
        e.arg2 = arg2;
        e.arg3 = arg3;
        e.arg4 = arg4;
        return e;
    }

    private static final ObjectPool<EventInfo, Void> sEventPool = new ObjectPool<EventInfo, Void>() {
        @Override
        protected EventInfo create(Void... arg) {
            return new EventInfo();
        }
    };

    public static class EventInfo extends Recycleable {
        public byte type;
        public int id;
        public int arg;
        public int arg2;
        public int arg3;
        public int arg5;
        public int arg4;
        public int arg6;
        public boolean flag;
        public Rect rect;
    }
}

File: app/src/main/java/org/monksanctum/xand11/comm/Event.kt
// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package org.monksanctum.xand11.comm

import android.graphics.Rect

import org.monksanctum.xand11.Client
import org.monksanctum.xand11.ObjectPool
import org.monksanctum.xand11.ObjectPool.Recycleable
import org.monksanctum.xand11.Utils

import java.lang.reflect.Field
import java.lang.reflect.Modifier

object Event {
    val ERROR: Byte = 0
    val REPLY: Byte = 1
    val KEY_PRESS: Byte = 2
    val KEY_RELEASE: Byte = 3
    val BUTTON_PRESS: Byte = 4
    val BUTTON_RELEASE: Byte = 5

    val MOTION_NOTIFY: Byte = 6
    val ENTER_NOTIFY: Byte = 7
    val LEAVE_NOTIFY: Byte = 8

    val FOCUS_IN: Byte = 9
    val FOCUS_OUT: Byte = 10

    val KEYMAP_NOTIFY: Byte = 11

    val EXPOSE: Byte = 12
    val GRAPHICS_EXPOSE: Byte = 13
    val NO_EXPOSE: Byte = 14
    val VISIBILITY_NOTIFY: Byte = 15
    val CREATE_NOTIFY: Byte = 16
    val DESTROY_NOTIFY: Byte = 17

    val UNMAP_NOTIFY: Byte = 18
    val MAP_NOTIFY: Byte = 19
    val MAP_REQUEST: Byte = 20
    val REPARENT_NOTIFY: Byte = 21

    val CONFIGURE_NOTIFY: Byte = 22
    val CONFIGURE_REQUEST: Byte = 23
    val GRAVITY_NOTIFY: Byte = 24
    val RESIZE_REQUEST: Byte = 25
    val CIRCULATE_NOTIFY: Byte = 26
    val CILCULATE_REQUEST: Byte = 27

    val PROPERTY_NOTIFY: Byte = 28
    val SELECTION_CLEAR: Byte = 29
    val SELECTION_REQUEST: Byte = 30
    val SELECTION_NOTIFY: Byte = 31
    val COLORMAP_NOTIFY: Byte = 32

    val CLIENT_MESSAGE: Byte = 33
    val MAPPING_NOTIFY: Byte = 34

    private val sNames = arrayOfNulls<String>(256)

    // TODO: Probobly should be per client or something...?
    private val sEvent = 1

    private val sEventPool = object : ObjectPool<EventInfo, Void>() {
        override fun create(vararg arg: Void): EventInfo {
            return EventInfo()
        }
    }

    fun sendPropertyChange(client: Client, window: Int, atom: Int, deleted: Boolean) {
        val writer = PacketWriter(client.clientListener.writer)

        writer.writeCard32(window)
        writer.writeCard32(atom)
        writer.writeCard32(client.timestamp)
        writer.writeByte((if (deleted) 1 else 0).toByte())
        writer.writePadding(15)
        client.clientListener.sendPacket(MAP_NOTIFY, writer)
    }

    fun sendMapNotify(client: Client, window: Int, event: Int,
                      overrideRedirect: Boolean) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.writeCard32(event)
        writer.writeCard32(window)
        writer.writeByte((if (overrideRedirect) 1 else 0).toByte())
        writer.writePadding(19)
        client.clientListener.sendPacket(MAP_NOTIFY, writer)
    }

    fun sendUnmapNotify(client: Client, window: Int, event: Int,
                        fromConfigure: Boolean) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.writeCard32(event)
        writer.writeCard32(window)
        writer.writeByte((if (fromConfigure) 1 else 0).toByte())
        writer.writePadding(19)
        client.clientListener.sendPacket(UNMAP_NOTIFY, writer)
    }

    fun sendSelectionRequest(client: Client, timestamp: Int, owner: Int, requestor: Int,
                             selection: Int, target: Int, property: Int) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.writeCard32(timestamp)
        writer.writeCard32(owner)
        writer.writeCard32(requestor)
        writer.writeCard32(selection)
        writer.writeCard32(target)
        writer.writeCard32(property)
        writer.writePadding(4)
        client.clientListener.sendPacket(SELECTION_REQUEST, writer)
    }

    fun sendSelectionNotify(client: Client, timestamp: Int, requestor: Int,
                            selection: Int, target: Int, property: Int) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.writeCard32(timestamp)
        writer.writeCard32(requestor)
        writer.writeCard32(selection)
        writer.writeCard32(target)
        writer.writeCard32(property)
        writer.writePadding(8)
        client.clientListener.sendPacket(SELECTION_NOTIFY, writer)
    }

    fun sendSelectionClear(client: Client, timestamp: Int, owner: Int, selection: Int) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.writeCard32(timestamp)
        writer.writeCard32(owner)
        writer.writeCard32(selection)
        writer.writePadding(16)
        client.clientListener.sendPacket(SELECTION_CLEAR, writer)
    }

    fun sendExpose(client: Client, window: Int, bounds: Rect) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.writeCard32(window)
        writer.writeCard16(bounds.left)
        writer.writeCard16(bounds.top)
        writer.writeCard16(bounds.width())
        writer.writeCard16(bounds.height())
        writer.writeCard16(0) // What?
        writer.writePadding(14)

        Utils.sBgHandler.post { client.clientListener.sendPacket(EXPOSE, writer) }
    }

    fun sendConfigureWindow(client: Client, event: Int, window: Int, sibling: Int,
                            bounds: Rect, borderWidth: Int, overrideDirect: Boolean) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.writeCard32(event)
        writer.writeCard32(window)
        writer.writeCard32(sibling)
        writer.writeCard16(bounds.left)
        writer.writeCard16(bounds.top)
        writer.writeCard16(bounds.width())
        writer.writeCard16(bounds.height())
        writer.writeCard16(borderWidth)
        writer.writeByte((if (overrideDirect) 1 else 0).toByte())
        writer.writePadding(5)

        client.clientListener.sendPacket(CONFIGURE_NOTIFY, writer)
    }

    fun sendKeyDown(client: Client, id: Int, x: Int, y: Int, keyCode: Int, state: Int) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.minorOpCode = keyCode.toByte()
        writer.writeCard32(client.timestamp)
        writer.writeCard32(3)
        writer.writeCard32(id)
        writer.writeCard32(0) // Child
        writer.writeCard16(x) // Root-x
        writer.writeCard16(y) // Root-y
        writer.writeCard16(x)
        writer.writeCard16(y)

        writer.writeCard16(state) // Keybutton state
        writer.writeByte(1.toByte()) // Same screen.
        writer.writePadding(1)

        client.clientListener.sendPacket(KEY_PRESS, writer)
    }

    fun sendKeyUp(client: Client, id: Int, x: Int, y: Int, keyCode: Int, state: Int) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.minorOpCode = keyCode.toByte()
        writer.writeCard32(client.timestamp)
        writer.writeCard32(3)
        writer.writeCard32(id)
        writer.writeCard32(0) // Child
        writer.writeCard16(0) // Root-x
        writer.writeCard16(0) // Root-y
        writer.writeCard16(x)
        writer.writeCard16(y)

        writer.writeCard16(state) // Keybutton state
        writer.writeByte(1.toByte())
        writer.writePadding(1)

        client.clientListener.sendPacket(KEY_RELEASE, writer)
    }

    fun sendEnter(client: Client, id: Int, x: Int, y: Int) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.minorOpCode = 1.toByte()

        writer.writeCard32(client.timestamp)
        writer.writeCard32(id)
        writer.writeCard32(id)
        writer.writeCard32(0)

        writer.writeCard16(0) // Root-x
        writer.writeCard16(0) // Root-y
        writer.writeCard16(x)
        writer.writeCard16(y)
        writer.writeCard16(0) // Keybutton state

        writer.writeByte(0.toByte()) // Normal, not grab
        writer.writeByte(3.toByte()) // Same screen, focus

        client.clientListener.sendPacket(ENTER_NOTIFY, writer)
    }

    fun sendLeave(client: Client, id: Int, x: Int, y: Int) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.minorOpCode = 1.toByte()

        writer.writeCard32(client.timestamp)
        writer.writeCard32(id)
        writer.writeCard32(id)
        writer.writeCard32(0)

        writer.writeCard16(0) // Root-x
        writer.writeCard16(0) // Root-y
        writer.writeCard16(x)
        writer.writeCard16(y)
        writer.writeCard16(0) // Keybutton state

        writer.writeByte(0.toByte()) // Normal, not grab
        writer.writeByte(3.toByte()) // Same screen, focus

        client.clientListener.sendPacket(LEAVE_NOTIFY, writer)
    }

    fun sendDestroy(client: Client, id: Int, arg: Int) {
        val writer = PacketWriter(client.clientListener.writer)
        writer.writeCard32(id)
        writer.writeCard32(arg)
        writer.writePadding(20)
        client.clientListener.sendPacket(DESTROY_NOTIFY, writer)
    }

    fun sendCreate(client: Client, id: Int, child: Int, x: Int, y: Int, width: Int,
                   height: Int, borderWidth: Int, overrideRedirect: Boolean) {
        val writer = PacketWriter(client.clientListener.writer)

        writer.writeCard32(id)
        writer.writeCard32(child)
        writer.writeCard16(x)
        writer.writeCard16(y)
        writer.writeCard16(width)
        writer.writeCard16(height)
        writer.writeCard16(borderWidth)
        writer.writeByte((if (overrideRedirect) 1 else 0).toByte())
        writer.writePadding(9)

        client.clientListener.sendPacket(CREATE_NOTIFY, writer)
    }

    fun populateNames() {
        val cls = Event::class.java
        for (field in cls!!.getDeclaredFields()) {
            if (Modifier.isStatic(field.getModifiers()) && field.getType() == Byte::class.javaPrimitiveType) {
                try {
                    val index = field.get(null) as Byte
                    sNames[index.toInt()] = field.name
                } catch (e: IllegalAccessException) {
                }

            }
        }
    }

    fun getName(what: Int): String? {
        return if (sNames[what] == null) {
            "Unknown($what)"
        } else sNames[what]
    }

    fun obtainInfo(type: Byte, id: Int, arg: Int): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.arg = arg
        return e
    }

    fun obtainInfo(type: Byte, id: Int, arg: Int, flag: Boolean): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.arg = arg
        e.flag = flag
        return e
    }

    fun obtainInfo(type: Byte, id: Int, arg: Int, arg2: Int, arg3: Int, arg4: Int,
                   arg5: Int, arg6: Int, flag: Boolean): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.arg = arg
        e.arg2 = arg2
        e.arg3 = arg3
        e.arg4 = arg4
        e.arg5 = arg5
        e.arg6 = arg6
        e.flag = flag
        return e
    }

    fun obtainInfo(type: Byte, id: Int, arg: Rect, flag: Boolean): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.rect = arg
        e.flag = flag
        return e
    }

    fun obtainInfo(type: Byte, id: Int, arg1: Int, arg2: Int, arg3: Int, rect: Rect,
                   flag: Boolean): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.arg = arg1
        e.arg2 = arg2
        e.arg3 = arg3
        e.rect = rect
        e.flag = flag
        return e
    }

    fun obtainInfo(type: Byte, id: Int, arg1: Int, arg2: Int, arg3: Int, arg4: Int,
                   arg5: Int): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.arg = arg1
        e.arg2 = arg2
        e.arg3 = arg3
        e.arg4 = arg4
        e.arg5 = arg5
        return e
    }

    fun obtainInfo(type: Byte, id: Int, arg1: Int, arg2: Int): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.arg = arg1
        e.arg2 = arg2
        return e
    }

    fun obtainInfo(type: Byte, id: Int, arg1: Int, arg2: Int, arg3: Int): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.arg = arg1
        e.arg2 = arg2
        e.arg3 = arg3
        return e
    }

    fun obtainInfo(type: Byte, id: Int, arg1: Int, arg2: Int, arg3: Int, arg4: Int): EventInfo {
        val e = sEventPool.obtain()
        e.type = type
        e.id = id
        e.arg = arg1
        e.arg2 = arg2
        e.arg3 = arg3
        e.arg4 = arg4
        return e
    }

    class EventInfo : Recycleable() {
        var type: Byte = 0
        var id: Int = 0
        var arg: Int = 0
        var arg2: Int = 0
        var arg3: Int = 0
        var arg5: Int = 0
        var arg4: Int = 0
        var arg6: Int = 0
        var flag: Boolean = false
        var rect: Rect? = null
    }
}
